mistralrs-core/src/pipeline/gguf_tokenizer.rs

```diff
@@ -136,6 +91,160 @@ pub fn convert_ggml_to_hf_tokenizer(content: &Content) -> Result<ConversionResul
     })
 }
 
+// TODO: Add support for additional tokenizer models: BPE, WordPiece, WordLevel
+// https://docs.rs/tokenizers/latest/tokenizers/models/enum.ModelWrapper.html
+#[derive(Debug)]
+enum TokenizerKind {
+    Unigram,
+}
+
+fn unigram_tokenizer(p: &PropsGGUF) -> Result<(Tokenizer, TokenizerKind, Vec<String>)> {
+    let PropsGGUF { unk, eos, bos, .. } = *p;
+    // Unigram (SentencePiece) default UNK is 0
+    let unk = unk.unwrap_or(0);
+
+    // Create the Tokenizer model:
+    let model = {
+        let vocab: Vec<(String, f64)> = {
+            let Some(s) = p.scores.as_ref() else {
+                anyhow::bail!(
+                    "`llama` unigram tokenizer is missing required metadata `tokenizer.ggml.scores`"
+                );
+            };
+            let scores = s.iter().cloned().map(|f_32| f_32 as f64);
+
+            p.tokens.iter().cloned().zip(scores).collect()
+        };
+
+        Unigram::from(vocab, Some(unk as usize), true).map_err(anyhow::Error::msg)?
+    };
+
+    let decoder = Decoder::Sequence(vec![
+        Decoder::Replace("▁", " "),
+        Decoder::ByteFallback,
+        Decoder::Fuse,
+        Decoder::Strip(' ', 1, 0),
+    ]);
+
+    let normalizer = Normalizer::Sequence(vec![
+        Normalizer::Prepend("▁"),
+        Normalizer::Replace(" ", "▁"),
+    ]);
+
+    let mut tokenizer: Tokenizer = TokenizerX::try_builder()
+        .with_model(model)
+        .with_decoder(decoder)
+        .with_normalizer(normalizer)
+        .build()?;
```

Comment:
This is the gist of the recent commits pushed. I tried to use the upstream `TokenizerBuilder` but it was problematic.

I've opened an issue for that upstream, but in the meantime have implemented `TokenizerX` struct with `buildstructor` to a similar builder API as upstream without the issues.

---

Is there a reference for this decoder + normalizer config for LLaMa?

I found several github discussions about the quirks with the [`Prepend` + `Replace` requirement](https://github.com/ggerganov/llama.cpp/pull/2398#issuecomment-1718177991) (_looks like `_` was the special character but it's not_), and the `Strip` is presumably to [remove the initial space inserted from the encode](https://github.com/ggerganov/llama.cpp/issues/167#issuecomment-1473369760)?:

> _replace occurrences of `▁` (Lower One Eighth Block) in a newly allocated string._

> _`SentencePiece` always encodes first token with whitespace even if you ask to prepend `<bos>` token._

I assume there is a spec somewhere to reference which could be added as a comment for additional context?

Reply:
I built that section of code by following a `tokenizer.json` for an HF sentencepiece model. I am not sure that there is a reference for these configs, but you can see that [Google's implementation](https://github.com/google/sentencepiece?tab=readme-ov-file#subword-regularization-and-bpe-dropout) also prepends this `_`.